### PR TITLE
Rework the timeout service(s)

### PR DESCRIPTION
**Pull request by gctucker**
Rework the implementation of the timeout / holdoff / closing services to address the following issues:
* retrieve only the nodes in the state relevant to the code, e.g. nodes in the 'done' state aren't needed for the timeout service
* split the implementation into different classes which run in different services to have better separation of the logic: Timeout for the main timeout, Holdoff for the holdoff timeout and Closing for nodes waiting in the Closing state
* avoid resubmitting the same nodes multiple times by first creating a set of nodes to update and then send them to the API
* work out the shortest time to sleep until next timeout / holdoff to allow longer polling periods and update nodes closer to the timeout / holdoff time

### CODE REVIEW

**Comment by gctucker** [comment]
Tested OK on staging, here's some final email reports:
```
kernelci/staging-stable staging-stable-20221125.0: 2 runs 1 failures

Summary
=======

Tree:     kernelci
Branch:   staging-stable
Describe: staging-stable-20221125.0
URL:      https://github.com/kernelci/linux.git
SHA1:     8e92708e80843a516e010a7781855b0de3a26c5e
Name            | Result   | Total    | Failures
----------------+----------+----------+---------
kunit           | pass     |      101 |        2
kver            | fail     |        0 |        0


Failing tests
=============

kunit
-----

* example.example_skip_test # SKIP this test should be skipped
* example.example_mark_skipped_test # SKIP this test should be skipped
```

```
kernelci/staging-next staging-next-20221125.0: 2 runs 0 failures

Summary
=======

Tree:     kernelci
Branch:   staging-next
Describe: staging-next-20221125.0
URL:      https://github.com/kernelci/linux.git
SHA1:     a2da84b86806322b81d094ccb938b426890ef2dd
Name            | Result   | Total    | Failures
----------------+----------+----------+---------
kver            | pass     |        0 |        0
kunit           | pass     |      254 |        0


Failing tests
=============
```

The `kver` test failed for stable, not sure why that's the case so it's something to investigate.

**Comment by gctucker** [comment]
Notifier log (3 kernel revisions being tested in parallel as staging jobs were broken for a while...):
```
[INFO] Listening for events... 
[INFO] Press Ctrl-C to stop.
[INFO] Time                        Commit        Node Id                   State      Result    Name
[INFO] 2022-11-25 12:12:12.360004  a2da84b86806  6380b11ce653ffd9ba8b58c1  Running    -         checkout
[INFO] 2022-11-25 12:12:13.136501  8e92708e8084  6380b11de653ffd9ba8b58c2  Running    -         checkout
[INFO] 2022-11-25 12:12:13.914916  08ad43d554ba  6380b11de653ffd9ba8b58c3  Running    -         checkout
[INFO] 2022-11-25 12:13:31.539139  a2da84b86806  6380b11ce653ffd9ba8b58c1  Available  -         checkout
[INFO] 2022-11-25 12:13:31.667538  a2da84b86806  6380b16be653ffd9ba8b58c4  Running    -         kver
[INFO] 2022-11-25 12:13:31.671866  a2da84b86806  6380b16be653ffd9ba8b58c5  Running    -         kunit
[INFO] 2022-11-25 12:13:55.311125  a2da84b86806  6380b16be653ffd9ba8b58c4  Done       Pass      kver
[INFO] 2022-11-25 12:14:52.744624  8e92708e8084  6380b11de653ffd9ba8b58c2  Available  -         checkout
[INFO] 2022-11-25 12:14:52.870636  8e92708e8084  6380b1bce653ffd9ba8b58c6  Running    -         kunit
[INFO] 2022-11-25 12:14:52.927460  8e92708e8084  6380b1bce653ffd9ba8b58c7  Running    -         kver
[INFO] 2022-11-25 12:15:06.685371  a2da84b86806  6380b16be653ffd9ba8b58c5  Done       Pass      kunit
[INFO] 2022-11-25 12:15:13.916184  8e92708e8084  6380b1bce653ffd9ba8b58c7  Done       Fail      kver
[INFO] 2022-11-25 12:16:09.194865  8e92708e8084  6380b1bce653ffd9ba8b58c6  Done       Pass      kunit
[INFO] 2022-11-25 12:16:11.195917  08ad43d554ba  6380b11de653ffd9ba8b58c3  Available  -         checkout
[INFO] 2022-11-25 12:16:11.313617  08ad43d554ba  6380b20be653ffd9ba8b5a2b  Running    -         kunit
[INFO] 2022-11-25 12:16:11.345159  08ad43d554ba  6380b20be653ffd9ba8b5a2c  Running    -         kver
[INFO] 2022-11-25 12:16:34.086306  08ad43d554ba  6380b20be653ffd9ba8b5a2c  Done       Pass      kver
[INFO] 2022-11-25 12:17:21.453999  08ad43d554ba  6380b20be653ffd9ba8b5a2b  Done       Pass      kunit
[INFO] 2022-11-25 12:23:32.055640  a2da84b86806  6380b11ce653ffd9ba8b58c1  Done       -         checkout
[INFO] 2022-11-25 12:24:53.272922  8e92708e8084  6380b11de653ffd9ba8b58c2  Done       -         checkout
[INFO] 2022-11-25 12:26:11.797348  08ad43d554ba  6380b11de653ffd9ba8b58c3  Done       -         checkout
```

**Comment by gctucker** [comment]
Timeout service log:
```
[INFO] Looking for nodes with lapsed timeout...
[INFO] Press Ctrl-C to stop.
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:59:58.908899
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:58:58.548735
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:58:59.334735
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:59:00.123735
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:59:00.897735
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:57:58.251020
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:57:59.826020
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:58:00.600020
[DEBUG] 6380b16be653ffd9ba8b58c5 timeout left: 5:59:18.377020
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:57:59.037020
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:56:57.892674
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:57:00.241674
[DEBUG] 6380b1bce653ffd9ba8b58c6 timeout left: 5:59:39.219674
[DEBUG] 6380b1bce653ffd9ba8b58c7 timeout left: 5:59:39.271674
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:56:58.678674
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:56:59.467674
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:55:57.524072
[DEBUG] 6380b20be653ffd9ba8b5a2b timeout left: 5:59:57.297072
[DEBUG] 6380b20be653ffd9ba8b5a2c timeout left: 5:59:57.326072
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:55:58.310072
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:55:59.099072
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:55:59.873072
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:54:57.228323
[DEBUG] 6380b20be653ffd9ba8b5a2b timeout left: 5:58:57.001323
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:54:58.014323
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:54:58.803323
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:54:59.577323
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:53:56.913617
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:53:57.699617
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:53:58.488617
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:53:59.262617
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:52:56.595990
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:52:57.381990
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:52:58.170990
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:52:58.944990
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:51:56.276769
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:51:57.062769
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:51:57.851769
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:51:58.625769
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:50:55.962557
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:50:56.748557
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:50:57.537557
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:50:58.311557
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:49:55.653347
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:49:56.439347
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:49:57.228347
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:49:58.002347
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:48:55.324815
[DEBUG] 6380b11ce653ffd9ba8b58c1 timeout left: 0:48:56.110815
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:48:56.899815
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:48:57.673815
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:47:55.021859
[DEBUG] 6380b11de653ffd9ba8b58c2 timeout left: 0:47:56.596859
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:47:57.370859
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:46:54.711916
[DEBUG] 6380b11de653ffd9ba8b58c3 timeout left: 0:46:57.060916
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:45:54.454689
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:44:54.148651
[DEBUG] 6380b11be653ffd9ba8b58c0 timeout left: 0:43:53.824144
```
No nodes reached timeout, the overall tarball timeout was set to 1h and holdoff to 10min.  All the tests were already done by the time the holdoff time was reached.

**Comment by gctucker** [comment]
Holdoff service log:
```
[INFO] Looking for nodes with lapsed holdoff...
[INFO] Press Ctrl-C to stop.
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:09:18.877963
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:08:18.683535
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:09:39.887535
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:07:18.528736
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:08:39.732736
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:09:58.189736
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:06:18.376724
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:07:39.580724
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:08:58.037724
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:05:18.230533
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:06:39.434533
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:07:57.891533
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:04:18.085456
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:05:39.289456
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:06:57.746456
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:03:17.936788
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:04:39.140788
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:05:57.597788
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:02:17.799852
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:03:39.003852
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:04:57.460852
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:01:17.621991
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:02:38.825991
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:03:57.282991
[DEBUG] 6380b11ce653ffd9ba8b58c1 holdoff left: 0:00:17.466820
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:01:38.670820
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:02:57.127820
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:01:21.098476
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:02:39.555476
[DEBUG] 6380b11ce653ffd9ba8b58c1 DONE
[DEBUG] 6380b11de653ffd9ba8b58c2 holdoff left: 0:00:20.525998
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:01:38.982998
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:01:18.358947
[DEBUG] 6380b11de653ffd9ba8b58c2 DONE
[DEBUG] 6380b11de653ffd9ba8b58c3 holdoff left: 0:00:17.551746
[DEBUG] 6380b11de653ffd9ba8b58c3 DONE
```
The 3 tarball nodes were transitioned to the 'done' state when the holdoff time was reached.

**Comment by gctucker** [comment]
The Closing service had no log (apart from the start-up message) as all the tests were complete before the tarball nodes reached their holdoff time.  Running longer tests such as fstests or reducing the holdoff window would cover this scenario.

I'll send a draft PR to reduce the holdoff time for tarball nodes as an experiment.  Ideally we should have end-to-end tests with temporary deployments to go through all these scenarios and validate all the state transitions against the documentation.

**Comment by gctucker** [comment]
With a holdoff time set to 1min (#182) and a single kernel revision this time:
```
[INFO] Listening for events... 
[INFO] Press Ctrl-C to stop.
[INFO] Time                        Commit        Node Id                   State      Result    Name
[INFO] 2022-11-25 12:46:07.683931  89d42c2b24d1  6380b90fe653ffd9ba8b5b21  Running    -         checkout
[INFO] 2022-11-25 12:47:02.346376  89d42c2b24d1  6380b90fe653ffd9ba8b5b21  Available  -         checkout
[INFO] 2022-11-25 12:47:02.480334  89d42c2b24d1  6380b946e653ffd9ba8b5b22  Running    -         kver
[INFO] 2022-11-25 12:47:02.482178  89d42c2b24d1  6380b946e653ffd9ba8b5b23  Running    -         kunit
[INFO] 2022-11-25 12:47:21.515374  89d42c2b24d1  6380b946e653ffd9ba8b5b22  Done       Pass      kver
[INFO] 2022-11-25 12:48:02.865101  89d42c2b24d1  6380b90fe653ffd9ba8b5b21  Closing    -         checkout
[INFO] 2022-11-25 12:48:09.142503  89d42c2b24d1  6380b946e653ffd9ba8b5b23  Done       Pass      kunit
[INFO] 2022-11-25 12:49:08.250416  89d42c2b24d1  6380b90fe653ffd9ba8b5b21  Done       -         checkout
```
```
[INFO] Looking for nodes with lapsed holdoff...
[INFO] Press Ctrl-C to stop.
[DEBUG] 6380b90fe653ffd9ba8b5b21 holdoff left: 0:00:54.891215
[DEBUG] 6380b90fe653ffd9ba8b5b21 HOLDOFF
```
```
[INFO] Looking for nodes that are done closing...
[INFO] Press Ctrl-C to stop.
[DEBUG] 6380b90fe653ffd9ba8b5b21 RUNNING: 1
[DEBUG] 6380b90fe653ffd9ba8b5b21 RUNNING: 0
[DEBUG] 6380b90fe653ffd9ba8b5b21 DONE
```

**Comment by gctucker** [comment]
Small problem:
```
2022-11-29T13:32:58.503142522Z 11/29/2022 01:32:58 PM UTC [INFO] Looking for nodes with lapsed holdoff...
2022-11-29T13:32:58.503202223Z 11/29/2022 01:32:58 PM UTC [INFO] Press Ctrl-C to stop.
2022-11-29T13:32:58.691652617Z 11/29/2022 01:32:58 PM UTC [ERROR] Traceback (most recent call last):
2022-11-29T13:32:58.691678817Z   File "/home/kernelci/pipeline/base.py", line 68, in run
2022-11-29T13:32:58.691689918Z     status = self._run(context)
2022-11-29T13:32:58.691694118Z   File "/home/kernelci/pipeline/timeout.py", line 138, in _run
2022-11-29T13:32:58.691701918Z     sleep_time = self._check_available_nodes(available_nodes)
2022-11-29T13:32:58.691710018Z   File "/home/kernelci/pipeline/timeout.py", line 112, in _check_available_nodes
2022-11-29T13:32:58.691714718Z     holdoff = datetime.fromisoformat(node['holdoff'])
2022-11-29T13:32:58.691719118Z TypeError: fromisoformat: argument must be str
```

File: src/timeout.py
Comment (mgalka) [inline src/timeout.py:180]: Maybe we could think of using `Enum` here, too?

**Review by mgalka — APPROVED** [review]

File: src/timeout.py
Comment (JenySadadia) [inline src/timeout.py:32]: How about declaring an instance variable with `['running', 'available', 'closing']` in `TimeoutService` class as the list is being used in multiple methods?
```
self._pending_node_states = ['running', 'available', 'closing']
```

File: src/timeout.py
Comment (JenySadadia) [inline src/timeout.py:81]: So, here polling period depends on the node's timeout value as well.
This way timed-out nodes will be updated more frequently than they would be in case of the actual polling period.
Won't it increase iterations?
What are the pros of doing this?

File: src/timeout.py
Comment (JenySadadia) [inline src/timeout.py:43]: I see you have used a different way of getting pending child nodes.
I used a simple way by calculating `all_child_nodes - completed_child_nodes`

**Review by JenySadadia — CHANGES_REQUESTED** [review]

File: src/timeout.py
Comment (gctucker) [reply src/timeout.py:180]: Yes, in fact I think this will need to be revisited as a follow-up as ideally we should be getting the list of states from the API code.  I'd rather leave it out of this PR for now as it's already quite big.

File: src/timeout.py
Comment (gctucker) [reply src/timeout.py:32]: Yes, see my reply to your other comment.

File: src/trigger.py
Comment (gctucker) [reply src/trigger.py:62]: Yes this should be in the YAML configuration once we have the ability to describe the pipeline jobs there.  We could add it as a command line argument in `trigger.py` in the meantime and remove it later.

File: src/timeout.py
Comment (gctucker) [reply src/timeout.py:32]: See my replies to @mgalka's comments, I think the definition of the nodes states should be revisited so I didn't want to optimise things here.  But yes we could add this, even as a class variable as it won't change with each instance.

File: src/timeout.py
Comment (gctucker) [reply src/timeout.py:81]: It just means, sleep for 60 seconds unless we know a node is going to timeout earlier than that.  It allows the sleep period to be longer as we can "wake up" in time for the next timeout.  This works really well with our current use-case as we have a handful of nodes to monitor and they normally don't time out.  It does mean that when a node times out or completes less than 60s before its timeout time, the query to get the pending nodes gets sent earlier as a side effect.  To scale better, we should look into subscribing to events about nodes entering the Available state and make use of the timeout timestamp when querying for pending nodes to retrieve only the ones that matter (I think I explained that in a comment somewhere).

File: src/timeout.py
Comment (gctucker) [reply src/timeout.py:43]: Yes, I know

File: src/trigger.py
Comment (gctucker) [reply src/trigger.py:62]: The same applies to the holdoff time set in `tarball.py`.  I think it's better to leave them as hard-coded so the command line argument doesn't become a feature that becomes difficult to remove later on.

File: src/timeout.py
Comment (gctucker) [reply src/timeout.py]: OK I've sent an update for that.

File: src/trigger.py
Comment (JenySadadia) [reply src/trigger.py:62]: That's true.

File: src/timeout.py
Comment (JenySadadia) [reply src/timeout.py:83]: Yes, after the PR https://github.com/kernelci/kernelci-api/pull/197 gets merged,
we can use filter {'timeout__lt': datetime.isoformat(datetime.utcnow())}

Comment (gctucker) [reply src/timeout.py:83]: Yes, I'll create some GitHub issues with some ideas for optimising the timeout implementation.

**Comment by JenySadadia** [comment]
What about setting node results on time out?
This whole block from the prior implementation is missing here:
```
def _set_node_result(self, node):
        """Set node result of timed out nodes
        If the node is in 'running' state, set the node result to 'incomplete'.
        When the timed out node is not in 'running' state, set the node result
        to 'fail' if one of the child nodes failed. Otherwise, set the result
        to 'pass'.
        """
        if node['state'] == 'running':
            node['result'] = 'incomplete'
            return node
        failed_child_nodes = self._db.count_nodes({
            'parent': node['_id'],
            'result': 'fail'
        })
        node['result'] = 'fail' if failed_child_nodes else 'pass'
        return 
```

**Comment by mgalka** [comment]
> @mgalka @JenySadadia I'll create some GitHub issues about potential improvements and optimisations (using Enum, using the `lt__` operator, subscribing to events...). Other than that, do you have any other comments on this PR?

I don't have any other comments. I just put those remarks for future and approved the PR

**Comment by gctucker** [comment]
> What about setting node results on time out? This whole block from the prior implementation is missing here:
> 
> ```
> def _set_node_result(self, node):
>         """Set node result of timed out nodes
>         If the node is in 'running' state, set the node result to 'incomplete'.
>         When the timed out node is not in 'running' state, set the node result
>         to 'fail' if one of the child nodes failed. Otherwise, set the result
>         to 'pass'.
>         """
>         if node['state'] == 'running':
>             node['result'] = 'incomplete'
>             return node
>         failed_child_nodes = self._db.count_nodes({
>             'parent': node['_id'],
>             'result': 'fail'
>         })
>         node['result'] = 'fail' if failed_child_nodes else 'pass'
>         return 
> ```

Yes maybe this should have been more explicit in the commit messages, ideally there should have been a commit to remove the code that was setting the result.

I think it's not the job of the timeout service to set the result, it should only be updating the state of the node.  In the code above, maybe some child nodes had already completed and their results were set to something, so overwriting them with `fail` doesn't seem right.  Generally speaking, the state machine described in the docs doesn't mention the result field but only the state field.  So I think it's better to stick to that and let the jobs set the result values.

**Comment by JenySadadia** [comment]
> > What about setting node results on time out? This whole block from the prior implementation is missing here:
> > ```
> > def _set_node_result(self, node):
> >         """Set node result of timed out nodes
> >         If the node is in 'running' state, set the node result to 'incomplete'.
> >         When the timed out node is not in 'running' state, set the node result
> >         to 'fail' if one of the child nodes failed. Otherwise, set the result
> >         to 'pass'.
> >         """
> >         if node['state'] == 'running':
> >             node['result'] = 'incomplete'
> >             return node
> >         failed_child_nodes = self._db.count_nodes({
> >             'parent': node['_id'],
> >             'result': 'fail'
> >         })
> >         node['result'] = 'fail' if failed_child_nodes else 'pass'
> >         return 
> > ```
> 
> Yes maybe this should have been more explicit in the commit messages, ideally there should have been a commit to remove the code that was setting the result.
> 
> I think it's not the job of the timeout service to set the result, it should only be updating the state of the node. In the code above, maybe some child nodes had already completed and their results were set to something, so overwriting them with `fail` doesn't seem right. Generally speaking, the state machine described in the docs doesn't mention the result field but only the state field. So I think it's better to stick to that and let the jobs set the result values.

Ideally, the node's result will not be set until it is in `done` state.
So when the service detects a timed-out node, it is setting the parent's result based on child nodes (set to `fail` if one of the child nodes failed otherwise `pass`). This is not overwriting of node results as the result of the timed-out node was not set earlier.
Jobs can only set the result when it is passed, failed, or skipped. They won't be aware of when the running node gets timed-out. How will the running node's result be set to `incomplete` when it times out if the `timeout` service won't do it?

**Comment by gctucker** [comment]
Ah yes sorry I misread  the code about failing child nides. I still think it's going one step too far to change the node result in the timeout service. A timeout doesn't need it's incomplete or that it failed, just that it's time to stop waiting for the result. What's the advantage of setting the result to "incomplete" upon timeout? If the job hasn't sent a result then there's just no result. This design docs don't mention anything about setting the result so it can be left empty. If we want to change that then we should review the design.

File: src/timeout.py
Comment (gctucker) [reply src/timeout.py]: See #191

**Comment by gctucker** [comment]
> @mgalka @JenySadadia I'll create some GitHub issues about potential improvements and optimisations (using Enum, using the `lt__` operator, subscribing to events...). Other than that, do you have any other comments on this PR?

See #192

**Comment by JenySadadia** [comment]
> Ah yes sorry I misread the code about failing child nides. I still think it's going one step too far to change the node result in the timeout service. A timeout doesn't need it's incomplete or that it failed, just that it's time to stop waiting for the result. What's the advantage of setting the result to "incomplete" upon timeout? If the job hasn't sent a result then there's just no result. This design docs don't mention anything about setting the result so it can be left empty. If we want to change that then we should review the design.

The advantage of setting the result to `incomplete` is to specifically identify that the particular node has been timed-out and the result will not be updated later.  But to set the node result after the timeout or not is still to be decided.
Although, the same can be conveyed by setting no result when the node moves to `done`.
So, it is fine to drop an `incomplete` result at the moment and we can add it later if it is required.

**Review by JenySadadia — APPROVED** [review]

**Comment by gctucker** [comment]
@mgalka @JenySadadia Do you need anything else before merging it?

**Comment by JenySadadia** [comment]
> @mgalka @JenySadadia Do you need anything else before merging it?

I think it is good to merge.

**Pull request merged**
